pkg/kp/rollstore/consul_store.go

```diff
@@ -0,0 +1,555 @@
+type sessionStore interface {
+	NewSession(name string, renewalCh <-chan time.Time) (kp.Session, chan error, error)
+}
+
+type consulStore struct {
+	kv KV
+
+	// Necessary for acquiring locks on the RCs referred to by an RU update
+	// request. This ensures that multiple rolling updates will not be
+	// admitted for the same replication controllers
+	rcstore rcstore.Store
+
+	// Needed for obtaining kp.Session
+	store sessionStore
+
+	// Used for the ability to specify RC for an RU to operate on using a
+	// label selector (see labelRCSpecifier)
+	labeler labels.Applicator
+
+	logger logging.Logger
+}
+
+var _ Store = consulStore{}
+
+func NewConsul(c *api.Client, logger *logging.Logger) Store {
+	if logger == nil {
+		logger = &logging.DefaultLogger
+	}
+	return consulStore{
+		kv:      c.KV(),
+		rcstore: rcstore.NewConsul(c, 3),
+		logger:  *logger,
+	}
+}
+
+func (s consulStore) Get(id roll_fields.ID) (roll_fields.Update, error) {
+	key, err := RollPath(id)
+	if err != nil {
+		return roll_fields.Update{}, nil
+	}
+
+	kvp, _, err := s.kv.Get(key, nil)
+	if err != nil {
+		return roll_fields.Update{}, consulutil.NewKVError("get", key, err)
+	}
+	if kvp == nil {
+		return roll_fields.Update{}, nil
+	}
+
+	return kvpToRU(kvp)
+}
+
+func (s consulStore) List() ([]roll_fields.Update, error) {
+	listed, _, err := s.kv.List(rollTree+"/", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make([]roll_fields.Update, 0, len(listed))
+	for _, kvp := range listed {
+		ru, err := kvpToRU(kvp)
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, ru)
+	}
+	return ret, nil
+}
+
+func (s consulStore) newRUCreationSession() (kp.Session, chan error, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, nil, util.Errorf("Could not determine hostname for RU create lock name: %s", err)
+	}
+
+	lockName := fmt.Sprintf(
+		"RU create lock for %s",
+		hostname,
+	)
+
+	return s.store.NewSession(lockName, nil)
+}
+
+// Obtains a lock for each RC in the list, or errors. RCs are locked in
+// lexicographical order by RC id to avoid deadlocks.
+func (s consulStore) lockRCs(rcIDs rc_fields.IDs, session kp.Session) error {
+	sort.Sort(rcIDs)
+	for _, rcID := range rcIDs {
+		_, err := s.rcstore.LockForUpdateCreation(rcID, session)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s consulStore) checkForConflictingUpdates(rcIDs rc_fields.IDs) error {
+	// Now that locks are held, check every RU and confirm that none of
+	// them refer to the new or old RCs
+	// This is potentially a scaling bottleneck (see function comment)
+	rus, err := s.List()
+	if err != nil {
+		return err
+	}
+
+	var conflictingID roll_fields.ID
+	for _, ru := range rus {
+		for _, id := range rcIDs {
+			if ru.NewRC == id || ru.OldRC == id {
+				conflictingID = ru.ID()
+				break
+			}
+		}
+	}
+
+	if conflictingID != "" {
+		return util.Errorf("RC %s conflicts with the request", conflictingID)
```

Comment:
This might be an error worth structuring more than just an error message, particularly since it will be shown to developers when deploy requests fail. It'd be nice to indicate which RU is causing the conflict, for instance.